src/shared/tpm2-util.c

```diff
@@ -2244,6 +2244,34 @@ int tpm2_digest_many_digests(
         return tpm2_digest_many(alg, digest, iovecs, n_data, extend);
 }
 
+/* This hashes the provided pin into a digest value, but also verifies that the final byte is not 0, because
+ * the TPM specification Part 1 ("Architecture") section Authorization Values (subsection "Authorization Size
+ * Convention") states "Trailing octets of zero are to be removed from any string before it is used as an
+ * authValue". Since the TPM doesn't know if the auth value is a "string" or just a hash digest, any hash
+ * digest that randomly happens to end in 0 must have the final 0 changed, or the TPM will remove it before
+ * using the value in its HMAC calculations, resulting in failed HMAC checks. */
+static int tpm2_get_pin_auth(TPMI_ALG_HASH hash, const char *pin, TPM2B_AUTH *ret_auth) {
+        TPM2B_AUTH auth = {};
+        int r;
+
+        assert(pin);
+        assert(ret_auth);
+
+        r = tpm2_digest_buffer(hash, &auth, pin, strlen(pin), /* extend= */ false);
+        if (r < 0)
+                return r;
+
+        assert(auth.size > 0);
+        if (auth.buffer[auth.size - 1] == 0) {
+                log_debug("authValue digest ends in 0 which the TPM will remove and cause HMAC authorization failures, adjusting.");
+                auth.buffer[auth.size - 1] = 0xff;
+        }
```

Comment:
hmm, i don't fully grok why this is necessary. can you explain?

I mean, as long as everyone drops the trailing NUL bytes we should be fine, no? or are you saying some code does and other code does not? can you be more specific on that?

Reply:
see https://github.com/systemd/systemd/issues/27716#issuecomment-1636750799

Reply:
Still not getting this. As long as the zeroes at the end are dropped systematically things should always result in the same state in the TPM, hence should be OK. 

Why specifically does that trigger any issue? can you elaborate for somebody a bit thick, like me?

Reply:
right, while the tpm drops the trailing 0, tpm2-tss doesn't, and so when tpm2-tss makes the hmac calculations they don't match what the tpm is expecting. If/when tpm2-tss is fixed, we could drop this, but of course we'd then need to set a minimum tpm2-tss version (that includes the fix).

@williamcroberts do you happen to know if there's an open bug in tpm2-tss for this?

Reply:
and also I'm not entirely sure it would be correct for tpm2-tss to remove (or alter) a trailing 0 from an authValue. What if the authValue is a hash digest? Maybe the tpm won't remove the trailing 0 in that case. And does tpm2-tss also mangle the TPM2B_SENSITIVE_CREATE authValue when creating a new object?

Probably the best outcome to this would be for a later revision of the spec to fix the ambiguity, but for now I don't think it's clear if tpm2-tss should modify trailing 0's or not.

Reply:
So the TPM stops processing auth values when it hits a NUL byte. This was a bug in the earlier implementation, but they kept the behavior around. For clarity, consider you had the string "passw\0rd", the auth value would get truncated to "passw".

Reply:
> ~So the TPM stops processing auth values when it hits a NUL byte. This was a bug in the earlier implementation, but they kept the behavior around. For clarity, consider you had the string "passw\0rd", the auth value would get truncated to "passw".~ Probably what I said is wrong, trying to dig that out of memory. @AndreasFuchsTPM might remember better.

Ahh I'm an idiot, I thought this was in the tpm2-tss code. Yeah their is no need to do this. The TPM will always trim the auth value on the ending NUL bytes... the TPM actually removes all trailing zeroes, this can be seen in:
MemoryRemoveTrailingZeros() within CheckPWAuthSession() in SessionProcess.c

Reply:
> > ~So the TPM stops processing auth values when it hits a NUL byte. This was a bug in the earlier implementation, but they kept the behavior around. For clarity, consider you had the string "passw\0rd", the auth value would get truncated to "passw".~ Probably what I said is wrong, trying to dig that out of memory. @AndreasFuchsTPM might remember better.
> 
> Ahh I'm an idiot, I thought this was in the tpm2-tss code. Yeah their is no need to do this. The TPM will always trim the auth value on the ending NUL bytes... the TPM actually removes all trailing zeroes, this can be seen in: MemoryRemoveTrailingZeros() within CheckPWAuthSession() in SessionProcess.c

right, but tpm2-tss *doesnt* trim those trailing null bytes, which is what causes the HMAC errors...

Reply:
Ah I see in the spec it does have a buried note that clarifies; in HMAC Computation under authValue there is "NOTE 2": `Trailing zeros are always removed from an authValue before it is used in an authorization computation.`

So yeah, this does seem like tpm2-tss should strip trailing 0's (probably in Esys_TR_SetAuth)?

Reply:
> > > ~So the TPM stops processing auth values when it hits a NUL byte. This was a bug in the earlier implementation, but they kept the behavior around. For clarity, consider you had the string "passw\0rd", the auth value would get truncated to "passw".~ Probably what I said is wrong, trying to dig that out of memory. @AndreasFuchsTPM might remember better.
> > 
> > 
> > Ahh I'm an idiot, I thought this was in the tpm2-tss code. Yeah their is no need to do this. The TPM will always trim the auth value on the ending NUL bytes... the TPM actually removes all trailing zeroes, this can be seen in: MemoryRemoveTrailingZeros() within CheckPWAuthSession() in SessionProcess.c
> 
> right, but tpm2-tss _doesnt_ trim those trailing null bytes, which is what causes the HMAC errors...

Ahh OK yeah I see that linked issue now, makes sense. I filed a bug in tpm2-tss: https://github.com/tpm2-software/tpm2-tss/issues/2664

Reply:
@ddstreet I can't reprod with the tpm2-tss stack, but I did file a [bug 2664](https://github.com/tpm2-software/tpm2-tss/issues/2664) and Juergen is looking into it on our side. However, can I nack this patch so it doesn't get released and could you respin a fix to address a passphrase ending in more than one zero byte (ie foopass\0\0\0\0\0\0\0), which is possible. It needs to rtrim all 0s.

Reply:
> @ddstreet I can't reprod with the tpm2-tss stack, but I did file a [bug 2664](https://github.com/tpm2-software/tpm2-tss/issues/2664) and Juergen is looking into it on our side. However, can I nack this patch so it doesn't get released and could you respin a fix to address a passphrase ending in more than one zero byte (ie foopass\0\0\0\0\0\0\0), which is possible. It needs to rtrim all 0s.

since we're using a hash digest, the actual text passphrase doesn't really matter, right?

And also we shouldn't need to handle any 0's other than the last one, since we're not actually trimming it; we are adjusting it (from 0->0xff). So any 0's before it should be handled normally. No?